### PR TITLE
Fault tolerance with redundancy and leader election

### PR DESCRIPTION
**Pull request by edwardalee**

### CODE REVIEW

**Comment by lhstrh** [comment]
This is a nice example, but I suggest we first fix the mentioned bugs before we feature it in the playground...

File: examples/C/src/leader-election/HeartbeatBully.lf [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+ * primary sends a heartbeat message once per second (by default). When the primary fails, a leader
+ * election protocol selects a new primary which then starts sending heartbeat messages. The program
+ * is set so that each primary fails after sending three heartbeat messages. When all nodes have
+ * failed, then the program exits.
+ *
+ * This example is designed to be run as a federated program with decentralized coordination.
+ * However, as of this writing, bugs in the federated code generator cause the program to fail
+ * because all federates get the same bank_index == 0. This may be related to these bugs:
+ *
+ * - https://github.com/lf-lang/lingua-franca/issues/1961
```
Comment (lhstrh) [inline examples/C/src/leader-election/HeartbeatBully.lf:21]: Both of these issues have now been closed.

Comment (edwardalee) [reply examples/C/src/leader-election/HeartbeatBully.lf:21]: The HeartbeatBully example now works federated!  Thanks all.

However, the NRP_LF.lf example still does not work. It fails with an STP violation.

Comment (lhstrh) [reply examples/C/src/leader-election/HeartbeatBully.lf:21]: Could you create an issue for this remaining problem in the `lingua-franca` repo? Thanks!

**Review by lhstrh — CHANGES_REQUESTED** [review]
These examples compile now, but there are still a bunch of FIXMEs in them that should be addressed before we merge this...

**Comment by edwardalee** [comment]
> These examples compile now, but there are still a bunch of FIXMEs in them that should be addressed before we merge this...

The FIXMEs are address in commit 39b70cbb5c1ba2fdf1a5bfa3adbcb2fd53725e8a.

**Pull request merged**